### PR TITLE
[JENKINS-56682] Reproduce and fix issue using types from shared libraries in Script initializers

### PR DESCRIPTION
**Pull request by dwnusbaum**
Same underlying problem as [JENKINS-56682](https://issues.jenkins-ci.org/browse/JENKINS-56682). See https://github.com/jenkinsci/script-security-plugin/pull/259 and https://github.com/jenkinsci/workflow-cps-plugin/pull/297.

This PR updates a bunch of dependencies since I had already had a wip commit to do that locally. More importantly, it adds a new test whose failing stack trace looks like: 

```
org.jenkinsci.plugins.scriptsecurity.sandbox.RejectedAccessException: Scripts not permitted to use new pkg.Foo
	at org.jenkinsci.plugins.scriptsecurity.sandbox.whitelists.StaticWhitelist.rejectNew(StaticWhitelist.java:271)
	at org.jenkinsci.plugins.scriptsecurity.sandbox.groovy.SandboxInterceptor.onNewInstance(SandboxInterceptor.java:174)
	at org.kohsuke.groovy.sandbox.impl.Checker$3.call(Checker.java:198)
	at org.kohsuke.groovy.sandbox.impl.Checker.checkedConstructor(Checker.java:203)
	at org.kohsuke.groovy.sandbox.impl.Checker$checkedConstructor.callStatic(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallStatic(CallSiteArray.java:56)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callStatic(AbstractCallSite.java:194)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callStatic(AbstractCallSite.java:214)
	at WorkflowScript.<init>(WorkflowScript:3)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at java.lang.Class.newInstance(Class.java:442)
	at org.codehaus.groovy.runtime.InvokerHelper.createScript(InvokerHelper.java:434)
Caused: groovy.lang.GroovyRuntimeException: Failed to create Script instance for class: class WorkflowScript. Reason
	at org.codehaus.groovy.runtime.InvokerHelper.createScript(InvokerHelper.java:466)
	at groovy.lang.GroovyShell.parse(GroovyShell.java:700)
	at org.jenkinsci.plugins.workflow.cps.CpsGroovyShell.doParse(CpsGroovyShell.java:133)
	at org.jenkinsci.plugins.workflow.cps.CpsGroovyShell.reparse(CpsGroovyShell.java:126)
	at org.jenkinsci.plugins.workflow.cps.CpsFlowExecution.parseScript(CpsFlowExecution.java:561)
	at org.jenkinsci.plugins.workflow.cps.CpsFlowExecution.start(CpsFlowExecution.java:522)
	at org.jenkinsci.plugins.workflow.job.WorkflowRun.run(WorkflowRun.java:320)
	at hudson.model.ResourceController.execute(ResourceController.java:97)
	at hudson.model.Executor.run(Executor.java:429)
```

EDIT: The PR now also pulls in a fix for the issue via workflow-cps and the test passes.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -64,17 +64,22 @@
...
         <changelist>-SNAPSHOT</changelist>
-        <jenkins.version>2.121.1</jenkins.version>
+        <jenkins.version>2.121.3</jenkins.version>
         <java.level>8</java.level>
         <workflow-cps-plugin.version>2.69</workflow-cps-plugin.version>
-        <git-plugin.version>3.6.4</git-plugin.version>
-        <scm-api-plugin.version>2.2.7</scm-api-plugin.version>
-        <workflow-scm-step-plugin.version>2.4</workflow-scm-step-plugin.version>
-        <workflow-multibranch-plugin.version>2.10</workflow-multibranch-plugin.version>
+        <git-plugin.version>3.9.3</git-plugin.version>
```
Comment (jglick) [inline pom.xml]: Note: will be some merge conflicts with #65.

File: pom.xml [hunk 2]
```diff
@@ -141,6 +146,11 @@
             <artifactId>commons-lang3</artifactId>
             <version>3.7</version>
         </dependency>
+        <dependency>
+            <groupId>org.jenkins-ci.plugins</groupId>
+            <artifactId>apache-httpcomponents-client-4-api</artifactId>
```
Comment (jglick) [inline pom.xml]: Surely this belongs in `dependencyManagement` not `dependencies`?

File: pom.xml [hunk 3]
```diff
@@ -64,17 +64,22 @@
     <properties>
         <revision>2.14</revision>
         <changelist>-SNAPSHOT</changelist>
-        <jenkins.version>2.121.1</jenkins.version>
+        <jenkins.version>2.121.3</jenkins.version>
         <java.level>8</java.level>
         <workflow-cps-plugin.version>2.69</workflow-cps-plugin.version>
```
Comment (jglick) [inline pom.xml]: Where in here are we picking up the new `script-security`? Or were we already picking it up?

File: src/test/java/org/jenkinsci/plugins/workflow/libs/LibraryAdderTest.java [hunk 4]
```diff
@@ -336,4 +337,22 @@
...
+        sampleRepo.git("add", "src");
+        sampleRepo.git("commit", "--message=init");
+        GlobalLibraries.get().setLibraries(Collections.singletonList(
+            new LibraryConfiguration("lib",
+                new SCMSourceRetriever(new GitSCMSource(null, sampleRepo.toString(), "", "*", "", true)))));
+        WorkflowJob p = r.jenkins.createProject(WorkflowJob.class, "p");
+        p.setDefinition(new CpsFlowDefinition(
+                "@Library('lib@master') import pkg.Foo\n" +
+                "import groovy.transform.Field\n" +
+                "@Field f = new Foo()\n", true));
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/libs/LibraryAdderTest.java]: Also check if something like this works:

```groovy
static def f = new Foo()
```

File: pom.xml (hunk 3 above)
Comment (dwnusbaum) [reply pom.xml]: We're using 1.58 after you updated it in https://github.com/jenkinsci/workflow-cps-global-lib-plugin/pull/67. I will go ahead and bump to 1.60 though.

File: src/test/java/org/jenkinsci/plugins/workflow/libs/LibraryAdderTest.java (hunk 4 above)
Comment (dwnusbaum) [reply src/test/java/org/jenkinsci/plugins/workflow/libs/LibraryAdderTest.java]: It looks like that works fine as-is.

Comment (jglick) [reply src/test/java/org/jenkinsci/plugins/workflow/libs/LibraryAdderTest.java]: Just fishing around for ways to convince the Groovy compiler to create an `<init>` or `<clinit>` method in `WorkflowScript`. The Java syntax

```java
{
  // …
}
```

presumably does not work in Groovy. Not sure if you can declare an explicit constructor

```groovy
def WorkflowScript() {
  new Foo()
}
```

etc.

Comment (dwnusbaum) [reply src/test/java/org/jenkinsci/plugins/workflow/libs/LibraryAdderTest.java]: Ok, it looks like if your Groovy script just declares a class, that gets used for the Script class, and you can use (static) initializer blocks as you would in Java, and they all throw similar errors as for `@Field` during `GroovyShell.parse`.

Error for a regular initializer block:
```
org.jenkinsci.plugins.scriptsecurity.sandbox.RejectedAccessException: Scripts not permitted to use new pkg.Foo
	at org.jenkinsci.plugins.scriptsecurity.sandbox.whitelists.StaticWhitelist.rejectNew(StaticWhitelist.java:271)
	at org.jenkinsci.plugins.scriptsecurity.sandbox.groovy.SandboxInterceptor.onNewInstance(SandboxInterceptor.java:174)
	at org.kohsuke.groovy.sandbox.impl.Checker$3.call(Checker.java:198)
	at org.kohsuke.groovy.sandbox.impl.Checker.checkedConstructor(Checker.java:203)
	at org.kohsuke.groovy.sandbox.impl.Checker$checkedConstructor.callStatic(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallStatic(CallSiteArray.java:56)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callStatic(AbstractCallSite.java:194)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callStatic(AbstractCallSite.java:214)
	at MyScript.<init>(WorkflowScript:3)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at java.lang.Class.newInstance(Class.java:442)
	at org.codehaus.groovy.runtime.InvokerHelper.createScript(InvokerHelper.java:438)
Caused: groovy.lang.GroovyRuntimeException: Failed to create Script instance for class: class MyScript. Reason
	at org.codehaus.groovy.runtime.InvokerHelper.createScript(InvokerHelper.java:466)
	at groovy.lang.GroovyShell.parse(GroovyShell.java:700)
```

Error for a static initializer block:

```
org.jenkinsci.plugins.scriptsecurity.sandbox.RejectedAccessException: Scripts not permitted to use new pkg.Foo
	at org.jenkinsci.plugins.scriptsecurity.sandbox.whitelists.StaticWhitelist.rejectNew(StaticWhitelist.java:271)
	at org.jenkinsci.plugins.scriptsecurity.sandbox.groovy.SandboxInterceptor.onNewInstance(SandboxInterceptor.java:174)
	at org.kohsuke.groovy.sandbox.impl.Checker$3.call(Checker.java:198)
	at org.kohsuke.groovy.sandbox.impl.Checker.checkedConstructor(Checker.java:203)
	at org.kohsuke.groovy.sandbox.impl.Checker$checkedConstructor.callStatic(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallStatic(CallSiteArray.java:56)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callStatic(AbstractCallSite.java:194)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callStatic(AbstractCallSite.java:214)
	at MyScript.<clinit>(WorkflowScript:3)
Caused: java.lang.ExceptionInInitializerError
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at java.lang.Class.newInstance(Class.java:442)
	at org.codehaus.groovy.runtime.InvokerHelper.createScript(InvokerHelper.java:438)
	at groovy.lang.GroovyShell.parse(GroovyShell.java:700)
```

Comment (jglick) [reply src/test/java/org/jenkinsci/plugins/workflow/libs/LibraryAdderTest.java]: > `at MyScript.<init>(WorkflowScript:3)`

Hmm, a corner case. See [JENKINS-31838](https://issues.jenkins-ci.org/browse/JENKINS-31838).

File: pom.xml (hunk 3 above)
Comment (dwnusbaum) [reply pom.xml]: Done in f9010950dd71060542d8b4dd61267a17ab6342df.

File: pom.xml (hunk 2 above)
Comment (dwnusbaum) [reply pom.xml]: Done in f9010950dd71060542d8b4dd61267a17ab6342df.

**Comment by dwnusbaum** [comment]
Ok, the new test should pass with after the update to workflow-cps in 83d1bea.

File: pom.xml [hunk 5]
```diff
@@ -64,17 +64,22 @@
...
-        <jenkins.version>2.121.1</jenkins.version>
+        <jenkins.version>2.121.3</jenkins.version>
         <java.level>8</java.level>
-        <workflow-cps-plugin.version>2.69</workflow-cps-plugin.version>
-        <git-plugin.version>3.6.4</git-plugin.version>
-        <scm-api-plugin.version>2.2.7</scm-api-plugin.version>
-        <workflow-scm-step-plugin.version>2.4</workflow-scm-step-plugin.version>
-        <workflow-multibranch-plugin.version>2.10</workflow-multibranch-plugin.version>
-        <workflow-step-api-plugin.version>2.19</workflow-step-api-plugin.version>
+        <workflow-cps-plugin.version>2.71-rc2340.ea3a8e1ac84d</workflow-cps-plugin.version>
```
Comment (jglick) [inline pom.xml]:
```suggestion
        <workflow-cps-plugin.version>2.71-rc2340.ea3a8e1ac84d</workflow-cps-plugin.version> <!-- TODO https://github.com/jenkinsci/workflow-cps-plugin/pull/297 -->
```

File: pom.xml (hunk 1 above)
Comment (jglick) [reply pom.xml]: Needs to be resolved.

Comment (dwnusbaum) [reply pom.xml]: Done in aadda23.

File: pom.xml (hunk 5 above)
Comment (dwnusbaum) [reply pom.xml]: Yeah, I need to do better about remembering TODO comments for this kind of thing. I updated to the released version in d26fa6c.
